In [2]:
# base imports
import pandas as pd
import numpy as np

In [3]:
# Data preparation: 
# Some data is missing (but we know what it should be) or some data should be changed.
df = pd.read_csv('./speed_dating_data.csv', encoding = 'ISO-8859-1')

def fooCareer(col):
    selected_df = df[col]
    career_c = {
        'lawyer': 1.,
        'law': 1.,
        'Economics': 7.,
        'tech professional': 5.
    }
    
    for x, row in selected_df.iterrows():
            career = df.at[x, col[0]]
            if career in career_c:
                df.at[x, col[1]] = career_c[career]
                

def fooGoal(col):
    selected_df = df[col]
    goal_c = {
        1: 1,
        2: 1,
        3: 1,
        4: 0,
        5: 1,
        6: 1,
    }
    
    for x, row in selected_df.iteritems():
        goal = df.at[x, col]
        if goal in goal_c:
            df.at[x, col] = goal_c[goal]

fooCareer(['career','career_c'])
fooGoal('goal')
df.to_csv('base_dating_data.csv', index=False)

In [4]:
# data is inconsistent:
# Some values (it’s based on waves) are 1-10 and others are 1-100 so
# they should be normalized so that all are 1-100.
df = pd.read_csv('base_dating_data.csv')

def func(lista):
    selected_df = df[lista]
    for x, row in selected_df.iterrows():
        multiplier = 100/row.sum()
        for col in lista:
            df.at[x, col] *= multiplier

lista = ['attr1_1', 'sinc1_1', 'intel1_1', 'fun1_1', 'amb1_1', 'shar1_1']
lista1 = ['attr_o', 'sinc_o', 'intel_o', 'fun_o', 'amb_o', 'shar_o']
lista3 = ['attr2_1', 'sinc2_1', 'intel2_1', 'fun2_1', 'amb2_1', 'shar2_1']
lista4 = ['attr1_2', 'sinc1_2', 'intel1_2', 'fun1_2', 'amb1_2', 'shar1_2']

func(lista)
func(lista1)
func(lista3)
func(lista4)

df.to_csv('base_dating_data.csv', index=False)

<ipython-input-4-023d9bb5a582>:9: RuntimeWarning: divide by zero encountered in double_scalars
  multiplier = 100/row.sum()
<ipython-input-4-023d9bb5a582>:11: RuntimeWarning: invalid value encountered in double_scalars
  df.at[x, col] *= multiplier


In [5]:
# Column met_o had values 1/2. It needs to be changed to 0/1.
df = pd.read_csv('./base_dating_data.csv')
df['met_o'] = df['met_o'].apply(lambda x: x - 1)
df.to_csv('base_dating_data.csv', index=False)

We removed some rows (by hand) because they didn't have `pid`.

### Below you can see code that prepares data for the basic matrix factorization.
Here in the base matrix we only have information about match. So we need a set of vectors, where each vector describes each date (holds both ids and information about match).

Task will requite two base matrices.
1. Matrix where men are "users" and women are "products". It will then be used to recommend women to men because matrix will say what's the predicted rating of a woman in eyes of man. Basically it will answer the question: **"How likely is that a man will like a woman?"**. Let's call this matrix/data frame **"men_like_women"**.
1. Matrix where women are "users" and men are "products". It will then be used to recommend men to women because matrix will say what's the predicted rating of a man in eyes of woman. Basically it will answer the question: **"How likely is that a woman will like a man?"**. Let's call this matrix/data frame **"women_like_men"**.

Why such analogies? It may help to understand how do this human relations task translates into recommender systems world.

These matrices will be used to train models (with matrix factorization) which will then be saved into csv files.

In [6]:
# Split into vectors, let's have two matrices as describes above.
base_df = pd.read_csv('./data_for_matrix.csv')
men_like_women_data = []
women_like_men_data = []

for _, row in base_df.iterrows():
    vector = {
        'id': row['iid'],
        'pid': row['pid'],
        'decision': row['dec'],
    }
    if row['gender'] == 0:
        # it's a woman
        women_like_men_data.append(vector)
    else:
        men_like_women_data.append(vector)

men_like_women_df = pd.DataFrame(men_like_women_data)
women_like_men_df = pd.DataFrame(women_like_men_data)

print('\n\nmen_like_women_df:')
print(men_like_women_df)
print('\nwomen_like_men_df:')
print(women_like_men_df)



men_like_women_df:
         id    pid  decision
0      11.0    1.0       0.0
1      11.0    2.0       0.0
2      11.0    3.0       0.0
3      11.0    4.0       0.0
4      11.0    5.0       0.0
...     ...    ...       ...
4179  552.0  526.0       0.0
4180  552.0  527.0       0.0
4181  552.0  528.0       0.0
4182  552.0  529.0       0.0
4183  552.0  530.0       0.0

[4184 rows x 3 columns]

women_like_men_df:
         id    pid  decision
0       1.0   11.0       1.0
1       1.0   12.0       1.0
2       1.0   13.0       1.0
3       1.0   14.0       1.0
4       1.0   15.0       1.0
...     ...    ...       ...
4179  530.0  548.0       0.0
4180  530.0  549.0       1.0
4181  530.0  550.0       0.0
4182  530.0  551.0       0.0
4183  530.0  552.0       1.0

[4184 rows x 3 columns]


## Let's prepare train and test data sets

### Make ids contiguous
We need to change the data. As you can see ids are not contiguous and they must be. So let's change that.

In [7]:
def proc_col(column):
    """Encodes a pandas column with continous ids."""
    unique = column.unique()
    old_to_new = {o: i for i, o in enumerate(unique)}
    return np.array([old_to_new.get(x, -1) for x in column])


def encode_data(df):
    """Encodes data with continous person and partner ids."""
    df = df.copy()
    for col_name in ["id", "pid"]:
        col = proc_col(df[col_name])
        df[col_name] = col
        df = df[df[col_name] >= 0]
    return df

And now use these functions to make ids contiguous.

In [8]:
print('Before:')
print(men_like_women_df)
men_like_women_df = encode_data(men_like_women_df)
print('\nAfter:')
print(men_like_women_df)

women_like_men_df = encode_data(women_like_men_df)

Before:
         id    pid  decision
0      11.0    1.0       0.0
1      11.0    2.0       0.0
2      11.0    3.0       0.0
3      11.0    4.0       0.0
4      11.0    5.0       0.0
...     ...    ...       ...
4179  552.0  526.0       0.0
4180  552.0  527.0       0.0
4181  552.0  528.0       0.0
4182  552.0  529.0       0.0
4183  552.0  530.0       0.0

[4184 rows x 3 columns]

After:
       id  pid  decision
0       0    0       0.0
1       0    1       0.0
2       0    2       0.0
3       0    3       0.0
4       0    4       0.0
...   ...  ...       ...
4179  276  269       0.0
4180  276  270       0.0
4181  276  271       0.0
4182  276  272       0.0
4183  276  273       0.0

[4184 rows x 3 columns]


### Let's split into train and test data sets
Standard sklearn function `train_test_split` doesn't do a job here because both test and train data sets should include all people. So after using `train_test_split` we need to transition some people between sets to ensure that both sets are correct. The same goes for partners.

In [9]:
from sklearn.model_selection import train_test_split as tts

def train_test_split(df, test_size=0.2):
    # Use standard train_test_split.
    df_train, df_test = tts(df, test_size=test_size)
    no_train_unique = len(df_train.id.unique())
    no_test_unique = len(df_test.id.unique())
    
    # See what people are missing in the test set.
    diff = np.setdiff1d(df_train.id.unique(), df_test.id.unique())
    for id in diff:
        # For every missing person we need to exchange them for a
        # different one so that sets still have the same number of elements.
        person_to_send_to_test = df_train.loc[df_train.id == id].iloc[0]
        person_to_send_to_train = None
        # Choose some person from the test set to send to the train set.
        ids = df_test.id.unique()
        np.random.shuffle(ids)
        for test_id in ids:
            person = df_test.loc[df_test.id == test_id]
            if len(person.index) > 1:
                person_to_send_to_train = person.iloc[0]
                break
        if person_to_send_to_train is not None:
            # Remove people that transit from old sets.
            # .name holds the id of that row in the oryginal df.
            df_train = df_train.drop(person_to_send_to_test.name)
            df_test = df_test.drop(person_to_send_to_train.name)
            # Add new people to sets.
            df_train = pd.concat([df_train, person_to_send_to_train.to_frame().T], ignore_index=True)
            df_test = pd.concat([df_test, person_to_send_to_test.to_frame().T], ignore_index=True)
        else:
            raise Exception("Couldn't find any person from people to send from the test to the train.")
        
    # See what partners are missing in the test set.
    diff = np.setdiff1d(df_train.pid.unique(), df_test.pid.unique())
    for pid in diff:
        # For every missing partner we need to exchange them for a
        # different one so that sets still have the same number of elements.
        partner_to_send_to_test = df_train.loc[df_train.pid == pid].iloc[0]
        partner_to_send_to_train = None
        
        # Choose some partner from the test set to send to the train set.
        # Need to make sure that both sets will still have all the people.
        pids = df_test.pid.unique()
        np.random.shuffle(pids)
        for test_pid in pids:
            partner = df_test.loc[df_test.pid == test_pid]
            if len(partner.index) > 1:
                # Make sure we don't remove a person completely.
                id = partner.iloc[0].id
                person_qty = len(df_test.loc[df_test.id == id].index)
                if person_qty > 1:
                    partner_to_send_to_train = partner.iloc[0]
                    break
        if partner_to_send_to_train is not None:
            # Remove partners that transit from old sets.
            df_train = df_train.drop(partner_to_send_to_test.name)
            df_test = df_test.drop(partner_to_send_to_train.name)
            # Add new people to sets.
            df_train = pd.concat([df_train, partner_to_send_to_train.to_frame().T], ignore_index=True)
            df_test = pd.concat([df_test, partner_to_send_to_test.to_frame().T], ignore_index=True)
        else:
            raise Exception("Couldn't find any partner from partners to send from the test to the train.")
                    
    df_train = df_train.sort_values(by='id')
    df_test = df_test.sort_values(by='id')
    return df_train.reset_index(drop=True), df_test.sort_values(by='id').reset_index(drop=True)

### And now just get correct data

In [10]:
men_like_women_train_df, men_like_women_test_df = train_test_split(men_like_women_df, test_size=0.2)
men_like_women_no_men = len(men_like_women_train_df.id.unique())
men_like_women_no_women = len(men_like_women_train_df.pid.unique())

women_like_men_train_df, women_like_men_test_df = train_test_split(women_like_men_df, test_size=0.2)
women_like_men_no_women = len(women_like_men_train_df.id.unique())
women_like_men_no_men = len(women_like_men_train_df.pid.unique())

## Let's create matrix factorization models

We will create and train several base MF (matrix factorization) models and for each of them let's do a cross validation to learn the best hyperparameters and parameters. Then we will compare the results and choose the best model.

Some general explanations for models:
* Models are train on only one batch because our data set is rather small.

Good reading resource: https://towardsdatascience.com/weight-initialization-techniques-in-neural-networks-26c649eb3b78

In [24]:
import torch
import torch.nn as nn
import torch.nn.functional as F

dev = torch.device('cpu')
#dev = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')
dev

device(type='cpu')

### Matrix factorization without bias

In [25]:
class MatrixFactorizationWithoutBiasNoXavier(nn.Module):
    def __init__(self, num_people, num_partners, weights=(0, 1), emb_size=100):
        super(MatrixFactorizationWithoutBiasNoXavier, self).__init__()
        self.person_emb = nn.Embedding(num_people, emb_size)
        self.partner_emb = nn.Embedding(num_partners, emb_size)
        self.person_emb.weight.data.uniform_(weights[0], weights[1])
        self.partner_emb.weight.data.uniform_(weights[0], weights[1])
        
    def forward(self, u, v):
        u = self.person_emb(u)
        v = self.partner_emb(v)
        # calculate dot product
        # u*v is a element wise vector multiplication
        return torch.sigmoid((u*v).sum(1))

    
class MatrixFactorizationWithoutBiasXavier(nn.Module):
    def __init__(self, num_people, num_partners, emb_size=100):
        super(MatrixFactorizationWithoutBiasXavier, self).__init__()
        self.person_emb = nn.Embedding(num_people, emb_size)
        self.partner_emb = nn.Embedding(num_partners, emb_size)
        torch.nn.init.xavier_uniform_(self.person_emb.weight)
        torch.nn.init.xavier_uniform_(self.partner_emb.weight)
        
    def forward(self, u, v):
        u = self.person_emb(u)
        v = self.partner_emb(v)
        # calculate dot product
        # u*v is a element wise vector multiplication
        return torch.sigmoid((u*v).sum(1))

    
# Example small models demonstrating weights
example_model_no_xavier = MatrixFactorizationWithoutBiasNoXavier(10, 10, emb_size=3)
example_model_xavier = MatrixFactorizationWithoutBiasXavier(10, 10, emb_size=3)
print("Model with without xavier weights are:\n")
for p in example_model_no_xavier.parameters():
    print(p)
print('\n\n', '='*20)
print("\n\nModel with with xavier weights are:\n")
for p in example_model_xavier.parameters():
    print(p)

Model with without xavier weights are:

Parameter containing:
tensor([[0.1749, 0.8508, 0.7031],
        [0.9645, 0.2625, 0.8845],
        [0.1108, 0.1730, 0.7423],
        [0.4509, 0.2567, 0.7866],
        [0.0333, 0.4978, 0.9542],
        [0.9384, 0.6168, 0.3752],
        [0.9995, 0.4324, 0.5110],
        [0.0040, 0.9172, 0.0996],
        [0.3266, 0.0691, 0.9855],
        [0.9825, 0.2714, 0.8465]], requires_grad=True)
Parameter containing:
tensor([[0.1496, 0.9583, 0.2665],
        [0.6169, 0.6255, 0.3359],
        [0.9838, 0.0149, 0.3768],
        [0.0933, 0.9356, 0.6598],
        [0.5312, 0.8991, 0.1143],
        [0.2122, 0.2495, 0.9514],
        [0.0941, 0.0507, 0.7865],
        [0.8398, 0.5725, 0.7058],
        [0.9414, 0.9717, 0.7890],
        [0.6272, 0.6888, 0.4022]], requires_grad=True)




Model with with xavier weights are:

Parameter containing:
tensor([[-0.5529, -0.0086, -0.6085],
        [ 0.0176,  0.2395,  0.2300],
        [ 0.2754, -0.0606, -0.2894],
        [-0.4241,  0

### Matrix factorization without bias

In [26]:
class MatrixFactorizationWithBiasNoXavier(nn.Module):
    def __init__(self, num_people, num_partners, weights=(0, 1), bias=(-0.01, 0.01), emb_size=100):
        super(MatrixFactorizationWithBiasNoXavier, self).__init__()
        self.person_emb = nn.Embedding(num_people, emb_size)
        self.person_bias = nn.Embedding(num_people, 1)
        self.partner_emb = nn.Embedding(num_partners, emb_size)
        self.parnter_bias = nn.Embedding(num_partners, 1)
        self.person_emb.weight.data.uniform_(weights[0], weights[1])
        self.partner_emb.weight.data.uniform_(weights[0], weights[1])
        self.person_bias.weight.data.uniform_(bias[0], bias[1])
        self.parnter_bias.weight.data.uniform_(bias[0], bias[1])
            
    def forward(self, u, v):
        u = self.person_emb(u)
        v = self.partner_emb(v)
        bias_u = self.person_bias(u).squeeze()
        bias_v = self.parnter_bias(v).squeeze()
        # calculate dot product
        # u*v is a element wise vector multiplication
        return torch.sigmoid((u*v).sum(1) + bias_u + bias_v)
    
    
class MatrixFactorizationWithBiasXavier(nn.Module):
    def __init__(self, num_people, num_partners, bias=(-0.01, 0.01), emb_size=100):
        super(MatrixFactorizationWithBiasXavier, self).__init__()
        self.person_emb = nn.Embedding(num_people, emb_size)
        self.person_bias = nn.Embedding(num_people, 1)
        self.partner_emb = nn.Embedding(num_partners, emb_size)
        self.parnter_bias = nn.Embedding(num_partners, 1)
        torch.nn.init.xavier_uniform_(self.person_emb.weight)
        torch.nn.init.xavier_uniform_(self.partner_emb.weight)
        self.person_bias.weight.data.uniform_(bias[0], bias[1])
        self.parnter_bias.weight.data.uniform_(bias[0], bias[1])
            
    def forward(self, u, v):
        u = self.person_emb(u)
        v = self.partner_emb(v)
        bias_u = self.person_bias(u).squeeze()
        bias_v = self.parnter_bias(v).squeeze()
        # calculate dot product
        # u*v is a element wise vector multiplication
        return torch.sigmoid((u*v).sum(1) + bias_u + bias_v)
    

# Example small models demonstrating weights
example_model_no_xavier = MatrixFactorizationWithBiasNoXavier(10, 10, bias=(0, 0), emb_size=3)
example_model_xavier = MatrixFactorizationWithBiasXavier(10, 10, emb_size=3)
print("Model with without xavier weights are:\n")
for p in example_model_no_xavier.parameters():
    print(p)
print('\n\n', '='*20)
print("\n\nModel with with xavier weights are:\n")
for p in example_model_xavier.parameters():
    print(p)

Model with without xavier weights are:

Parameter containing:
tensor([[0.2547, 0.1663, 0.2969],
        [0.1486, 0.1388, 0.0501],
        [0.0739, 0.0908, 0.9775],
        [0.3215, 0.4497, 0.6155],
        [0.2173, 0.9022, 0.3188],
        [0.9592, 0.8097, 0.1082],
        [0.2767, 0.9906, 0.6686],
        [0.5387, 0.1899, 0.2221],
        [0.0841, 0.8829, 0.0987],
        [0.6080, 0.3791, 0.1198]], requires_grad=True)
Parameter containing:
tensor([[0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.],
        [0.]], requires_grad=True)
Parameter containing:
tensor([[0.3689, 0.6824, 0.1335],
        [0.2672, 0.1401, 0.7044],
        [0.1378, 0.5778, 0.1191],
        [0.1191, 0.9690, 0.6474],
        [0.1747, 0.4584, 0.2007],
        [0.5672, 0.1445, 0.4801],
        [0.5076, 0.1795, 0.6687],
        [0.4809, 0.3441, 0.9742],
        [0.9235, 0.0237, 0.5959],
        [0.7514, 0.9405, 0.3674]], requires_grad=True)
Parameter c

### Training and testing functions are below

In [27]:
def test(model, df_test, verbose=False):
    model.eval()
    # .to(dev) puts code on either gpu or cpu.
    people = torch.LongTensor(df_test.id.values).to(dev)
    partners = torch.LongTensor(df_test.pid.values).to(dev)
    decision = torch.FloatTensor(df_test.decision.values).to(dev)
    y_hat = model(people, partners)
    loss = F.mse_loss(y_hat, decision)
    if verbose:
        print('test loss %.3f ' % loss.item())
    return loss.item()


# Default values assigned below are ones that I found online.
# Cross validadtion will be done later but it's good to have some defaults.
def train(model, df_train, epochs=100, learning_rate=0.01, weight_decay=1e-5, verbose=False):
    optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate, weight_decay=weight_decay)
    model.train()
    
    for epoch in range(epochs):
        # From numpy to PyTorch tensors.
        # .to(dev) puts code on either gpu or cpu.
        people = torch.LongTensor(df_train.id.values).to(dev)
        partners = torch.LongTensor(df_train.pid.values).to(dev)
        decision = torch.FloatTensor(df_train.decision.values).to(dev)
        
        # calls forward method of the model
        y_hat = model(people, partners)
        # Using mean squared errors loss function
        loss = F.mse_loss(y_hat, decision)
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        
        if verbose and epoch % 100 == 0: 
            print(loss.item())

# Let's finally train models and choose the best one

## Let's first train men_like_women

## Cross validation on each model

In [28]:
from itertools import product
from tqdm import tqdm

epochs_poss = list(range(10, 310, 10))
weight_decay_poss = [0.0, 1e-6, 1e-5, 1e-4, 0.0001, 0.001, 0.01, 0.1, 1.0]
learning_rate_poss = [1e-6, 1e-5, 1e-4, 0.0001, 0.001, 0.01, 0.1, 1.0]
weights_poss = [(0, 1), (-1, 1), (0, 0.2), (-0.2, 0.2), 'xavier']
bias_poss = [None, (0, 0), (-0.01, 0.01), (0, 1), (-1, 1)]
emb_size_poss = [10, 50, 100, 150]

def cross_vaild_model(model, train_df, test_df, verbose=False):
    """Function to choose the best hyperparameters for a model."""
    min_loss = float('inf')
    best_settings = None

    for (epochs, wd, lr) in tqdm(product(epochs_poss, weight_decay_poss, learning_rate_poss)):
        train(model, train_df, epochs=epochs, learning_rate=lr, weight_decay=wd)
        test_loss = test(model, test_df)
        if (test_loss < min_loss) or (test_loss == min_loss and best_settings is not None and epochs < best_settings['epochs']):
            min_loss = test_loss
            best_settings = {'epochs': epochs, 'weight_decay': wd, 'learning_rate_poss': lr}
    if verbose:
        print('min loss %.3f' % min_loss)
        print('best settings are', best_settings)
    return min_loss, best_settings


def cross_vaild(num_people, num_partners, train_df, test_df, verbose=False):
    """Function to choose the best model."""
    min_loss = float('inf')
    best_settings = None
    
    for (weights, bias, emb_size) in tqdm(product(weights_poss, bias_poss, emb_size_poss)):
        model = None
        if weights == 'xavier':
            if bias is None:
                model = MatrixFactorizationWithoutBiasXavier(num_people, num_partners, emb_size=emb_size).to(dev)
            else:
                model = MatrixFactorizationWithBiasXavier(num_people, num_partners, bias=bias, emb_size=emb_size).to(dev)
        else:
            if bias is None:
                model = MatrixFactorizationWithoutBiasNoXavier(num_people, num_partners, weights=weights, emb_size=emb_size).to(dev)
            else:
                model = MatrixFactorizationWithBiasNoXavier(num_people, num_partners, weights=weights, bias=bias, emb_size=emb_size).to(dev)
        
        model_min_loss, model_best_settings = cross_vaild_model(model, train_df, test_df, verbose)
        if (model_min_loss < min_loss) or (model_min_loss == min_loss and best_settings and emb_size < best_settings['emb_size']):
            min_loss = model_min_loss
            curr_settings = {'model': model, 'weights': weights, 'bias': bias, 'emb_size': emb_size}
            best_settings = {**curr_settings, **model_best_settings}
            
    return min_loss, best_settings

In [29]:
# model.eval()
# print(model(torch.LongTensor(men_like_women_train_df.id).to(dev), torch.LongTensor(men_like_women_train_df.pid).to(dev)))

## Let's create logistic regression
Now, to create regression models, both used before and after, we will modify data set in such ways:
1. With modified income column - here the missing values are filled with the mean of the column and modified attr_5_1 columns - here the missing values are filled with the mean of the columns.
2. Without income and attr_5_1 columns.

#### Later we will aply the logic to combine this with MF.

It's going to be used:
1. Before matrix factorization - output of this model will be used as a input to MF.
2. Before matrix factorization - output of MF will be used as input in this model, as one of the features.

We will try 3 diffetent combinations:
1. Regression before MF.
2. Regression after MF.
3. Regression before and after MF.


### Data preprocessing for logistic regression

In [30]:
def fill_with_mean(df):
    """Function fills missing values with mean of the column"""
    col_means = df.mean()
    for index, row in df.iterrows():
        for column in df.columns:
            if column != 'field' and column != 'race_o' and column != 'met_o' and column != 'race' and column != 'goal' and column != 'date' and column != 'go_out' and column != 'career_c':
                if np.isnan(row[column]):
                    df.at[index, column] = col_means[column]
            elif column == 'field' and np.isnan(row['field']):
                df.at[index, column] = 8
            elif column == 'race_o' and np.isnan(row['race_o']):
                df.at[index, column] = 2
            elif column == 'race' and np.isnan(row['race']):
                df.at[index, column] = 2
            elif column == 'goal' and np.isnan(row['goal']):
                df.at[index, column] = 1
            elif column == 'date' and np.isnan(row['date']):
                df.at[index, column] = 6
            elif column == 'go_out' and np.isnan(row['go_out']):
                df.at[index, column] = 2
            elif column == 'career_c' and np.isnan(row['career_c']):
                df.at[index, column] = 2
            elif column == 'met_o' and (np.isnan(row['met_o']) or row['met_o'] != 0):
                df.at[index, column] = 1
    return df

### Option 1
Modified income column - here the missing values are filled with the mean of the column and modified attr_5_1 columns - here the missing values are filled with the mean of the columns.

In [31]:
regression_df_1 = base_df.copy()
regression_df_1 = fill_with_mean(regression_df_1)

### Option 2
Without income and attr_5_1 columns.

In [32]:
regression_df_2 = base_df.copy()
cols_to_drop = ['income', 'attr5_1', 'sinc5_1', 'intel5_1', 'fun5_1', 'amb5_1']
regression_df_2 = regression_df_2.drop(columns=cols_to_drop)
regression_df_2 = fill_with_mean(regression_df_2)

### Definition of logistic regression

In [33]:
class LogisticRegression(torch.nn.Module):
    def __init__(self, input_dim, output_dim):
        super(LogisticRegression, self).__init__()
        #(out)input_dim is size of our (out)input data
        self.linear = torch.nn.Linear(input_dim, output_dim)

    def forward(self, x):
        pred_y = torch.sigmoid(self.linear(x))
        return torch.squeeze(pred_y, 1)

In [34]:
def train(model, x, y, epochs=100, learning_rate=0.01, weight_decay=1e-5, verbose=False):
    optimizer = torch.optim.Adam(model.parameters(), lr=learning_rate, weight_decay=weight_decay)
    model.train()
    
    for epoch in range(epochs):
        # From numpy to PyTorch tensors.
        # .to(dev) puts code on either gpu or cpu.
        X = torch.FloatTensor(x).to(dev)
        Y = torch.FloatTensor(y).to(dev)
        
        # calls forward method of the model
        y_hat = model(X)
        # Using mean squared errors loss function
        loss = F.mse_loss(y_hat, Y)
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()
        
        if verbose and epoch % 10 == 0: 
            print(loss.item())
            
def test(model, x, y, verbose=False):
    model.eval()
    # .to(dev) puts code on either gpu or cpu.
    X = torch.FloatTensor(x).to(dev)
    Y = torch.FloatTensor(y).to(dev)
    y_hat = model(X)
    # Using mean squared errors loss function
    loss = F.mse_loss(y_hat, Y)
    if verbose:
        print('test loss %.3f ' % loss.item())
    return loss.item()

### Let's find the best model using Cros validation:

In [35]:
epochs_poss = list(range(100, 510, 10))
weight_decay_poss = [0.0,  1e-8, 1e-7, 1e-6, 1e-5, 1e-4, 0.0001, 0.001, 0.01, 0.1, 1.0]
learning_rate_poss = [1e-8, 1e-7, 1e-6, 1e-5, 1e-4, 0.0001, 0.001, 0.01, 0.1, 1.0] 
#testing every combination of epoch ~ weight_decay ~ learning_rate values
def cross_valid_regression(model, train_x, train_y, test_x, test_y, verbose=False):
    """Function to choose the best hyperparameters for a model."""
    min_loss = np.inf
    best_settings = None

    for (epochs, wd, lr) in tqdm(product(epochs_poss, weight_decay_poss, learning_rate_poss)):
        train(model, train_x, train_y, epochs=epochs, learning_rate=lr, weight_decay=wd)
        test_loss = test(model, test_x, test_y)
        if (test_loss < min_loss) or (test_loss == min_loss and best_settings is not None and epochs < best_settings['epochs']):
            min_loss = test_loss
            best_settings = {'epochs': epochs, 'weight_decay': wd, 'learning_rate_poss': lr}
    if verbose:
        print('min loss %.3f' % min_loss)
        print('best settings are', best_settings)
    return min_loss, best_settings


In [36]:
#train_sample defines size of train set
train_sample = 0.8
#our output is only one number
output_dim = 1

#simple division into training and testing sets
msk = np.random.rand(len(regression_df_1)) < train_sample
train_set = regression_df_1[msk]
test_set = regression_df_1[~msk]

#we want to predict 'dec' basing on other attributes. We dont want 'iid' and 'pid' columns
#to have any impact on results
train_x = train_set.drop(['dec', 'iid', 'pid'], axis=1).values
train_y = train_set['dec'].values 

test_x = test_set.drop(['dec', 'iid', 'pid'], axis=1).values
test_y = test_set['dec'].values 

input_dim = train_x.shape[1]
model = LogisticRegression(input_dim, output_dim).to(dev)
print(cross_valid_regression(model, train_x, train_y, test_x, test_y))

4510it [37:16,  2.02it/s]

(0.23222902417182922, {'epochs': 460, 'weight_decay': 0.1, 'learning_rate_poss': 0.0001})


***After trying various different combinations of parameters we have found the best one:***
- epochs: 460
- weight_decay: 0.1
- learning_rate_poss: 0.0001

This combination resulted with loss ~ 0.232
